Whitelisted privileged containers admission controller

Makdaam opened the pull request:
```release-note
Enable use of whitelisted privileged containers on ARO service
```

Makdaam commented:
All elements are coded (not finished - no apps.Deployment coverage yet).
Currently investigating issues which didn't show up during manual deployment.

jim-minter commented on pkg/startup/v11/data/master/etc/aro-admission-controller/aro-admission-controller.yaml:
```diff
@@ -0,0 +1,5 @@
+whitelist:
+  - {{ .Config.Images.ControlPlane | quote }}
```
1) are these going to be regexes?  If so need ^ $ and a quotemeta
2) can the admission controller detect pods that are going to be scheduled to masters (via their nodeselector?) and ignore them, such that they need not be listed here?

jim-minter commented on pkg/startup/v11/data/master/etc/origin/master/master-config.yaml:
```diff
@@ -1,5 +1,10 @@
 admissionConfig:
   pluginConfig:
+    ValidatingAdmissionWebhook:
```
curious if you are sure that this is working? cf https://bugzilla.redhat.com/show_bug.cgi?id=1635918 ?

Makdaam replied:
Verified that the Validating admission webhook is enabled with this config and works. The creation of pods outside of the whitelist is blocked.

jim-minter commented on pluginconfig/pluginconfig-311.yaml:
```diff
@@ -50,6 +50,7 @@ versions:
       genevaTDAgent: osarpint.azurecr.io/acs/td-agent:master.20190228.1
       azureControllers: quay.io/openshift-on-azure/azure:v7.0
       canary: quay.io/openshift-on-azure/azure:v7.0
+      aroAdmissionController: placeholder
```
can these placeholder lines be removed? if not, why not?

Makdaam replied:
Removed.

jim-minter commented on pluginconfig/pluginconfig-311.yaml:
```diff
@@ -183,37 +187,38 @@ versions:
       genevaTDAgent: osarpint.azurecr.io/acs/td-agent:master.20190228.1
       azureControllers: quay.io/openshift-on-azure/azure:latest
       canary: quay.io/openshift-on-azure/azure:latest
+      aroAdmissionController: quay.io/makdaam/aro-admission-controller:init-dirty
```
(reminder: must fix before merge)

Makdaam replied:
Changed to the standard azure image + added AZURE_IMAGE env var override in fakerp.

jim-minter commented on pkg/startup/v11/data/master/etc/origin/node/pods/aroadmissioncontroller.yaml:
```diff
@@ -0,0 +1,70 @@
+          name: master-kube
+  hostNetwork: true
+  priorityClassName: system-node-critical
+  volumes:
```
are all of these really needed?

jim-minter commented on pkg/sync/v11/data/ClusterRole.rbac.authorization.k8s.io/privileged-creator.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: privileged-creator
```
what's this for?

Makdaam replied:
Moved to ClusterRole customer-admin-cluster.

Makdaam commented:
/retest e2e-create-20190430
cluster create timed out on waiting for sync pod, trying to investigate on a CI cluster since it works well on my locally created one.

Makdaam commented:
/test e2e-create-20190430

Makdaam commented:
/retest
turns out "waiting on sync pod" timeout can happen if the CI jobs run in the wrong order? I'll have to verify that.

Makdaam commented:
Some information which will help with reading the PR.
First commits until "sync and configuration set" only modify the sync pod and assume an admission controller is being build outside the openshift-azure repo.
"moved sync pod and startup configuration from v11 to v12, cleaned up" moves the changes from v11 to v12, after the v11 cut.
Starting with the "code moved, dependencies not updated" until the end it's mostly about adding the code of the actual Admission controller (vendor commit contains all the vendor changes).

The Admission Controller is run as a static pod on the masters, configured with a ValidatingWebhookConfiguration resource. The sync pod blocks during the first run until 3 instances of the Admission Controller pod are up before enabling it by adding the ValidatingWebhookConfiguration resource. This solves the chicken and egg problem of requiring AC to approve itself.

The relaxed customer-admin-cluster role is also applied after the AC is online.

jim-minter commented:
@Makdaam for review please rebase such that the vendor changes are in one commit and the remaining changes are in 1 or few commits.

Makdaam commented:
Squashed to 4 commits, moved the vendor commit to the end for ease of selecting diffs. To be chronologically accurate it should be before the "fixing dependencies" commit.

Also trying to fix issues with retrying round tripper, but the race condition doesn't trigger when I run the tests locally.

jim-minter commented on pkg/entrypoint/admissioncontroller/admissioncontroller_test.go:
```diff
@@ -0,0 +1,1084 @@
+		{
+			name: "protected SCC, added user, allow",
+			scc: `{
+				"metadata": {
```
+1 for the unit test, but its verbosity adds a maintenance overhead - can you find a way to simplify the boilerplate?

jim-minter commented on pkg/entrypoint/admissioncontroller/admissioncontroller_test.go:
```diff
@@ -0,0 +1,1084 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "default",
+			Annotations: map[string]string{
+				"openshift.io/sa.scc.uid-range": "1000/10",
```
is any of this needed here?

Makdaam replied:
No, removed.

jim-minter commented on pkg/entrypoint/admissioncontroller/admissioncontroller_test.go:
```diff
@@ -0,0 +1,1084 @@
+		t.Fatal(err)
+	}
+
+	reqNoUID, err := json.Marshal(&admissionv1beta1.AdmissionReview{
```
what is any of this actually testing?

Makdaam replied:
Malformed review requests. We have a separate code path for review requests with empty UID which leads to an error. I don't know if testing it is necessary.

jim-minter commented on pkg/entrypoint/admissioncontroller/podwhitelist.go:
```diff
@@ -0,0 +1,145 @@
+)
+
+func (ac *admissionController) handleWhitelist(w http.ResponseWriter, r *http.Request) {
+	unpackers := map[string]func(runtime.Object) (core.PodSpec, metav1.ObjectMeta, string){
```
simplify:
```
func unpack(o runtime.Object) (*core.Pod, error) {
	switch o := o.(type) {
	case *core.Pod:
		return o, nil
	case *extensions.DaemonSet:
		pod := &core.Pod{ObjectMeta: o.Spec.Template.ObjectMeta, Spec: o.Spec.Template.Spec}
		pod.Namespace = o.Namespace
		return pod, nil
	...
	default:
		return nil, fmt.Errorf("unimplemented gvk %v", o.GetObjectKind().GroupVersionKind())
	}
}
```

Makdaam replied:
Done.

jim-minter commented on pkg/entrypoint/admissioncontroller/podwhitelist.go:
```diff
@@ -0,0 +1,145 @@
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+	gvk := schema.GroupVersionKind{Group: req.Kind.Group, Version: req.Kind.Version, Kind: req.Kind.Kind}
```
think we discussed this before - do you really need this line?

Makdaam replied:
Turns out that it's needed because the embedded object doesn't have GroupVersionKind set, so the Decode function has to get that value from the review request.

jim-minter commented on pkg/entrypoint/admissioncontroller/podwhitelist.go:
```diff
@@ -0,0 +1,145 @@
+//checkPodSpec checks if the Pod spec is either whitelisted or will match the restricted scc, then prepares an HTTP response
+// interface{} is used to allow core.Pod from both the Openshift and Kubernetes APIs
+func (ac *admissionController) checkPodSpec(podSpec core.PodSpec, oMeta metav1.ObjectMeta, namespace string, w http.ResponseWriter, uid types.UID) {
+	pod := new(core.Pod)
```
this will boil away with the above changes to unpack()

Makdaam replied:
Done.

jim-minter commented on pkg/entrypoint/admissioncontroller/podwhitelist.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// podIsWhitelisted returns true if all images of all containers are whitelisted
+func (ac *admissionController) podSpecIsWhitelisted(spec *core.PodSpec) bool {
```
needs unit test

jim-minter commented on pkg/startup/v12/data/master/etc/origin/node/pods/aroadmissioncontroller.yaml:
```diff
@@ -0,0 +1,50 @@
+        path: /etc/aro-admission-controller
+      name: etc-adc
+    - hostPath:
+        path: /root/.kube
```
use /etc/origin/master/admin.kubeconfig

Makdaam replied:
Done.

jim-minter commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+
+	restconfig, err := kubeconfig.ClientConfig()
+	if err != nil {
+		restconfig, err = rest.InClusterConfig()
```
do you actually ever use InClusterConfig?  If not, remove

Makdaam replied:
Removed.

jim-minter commented on pkg/sync/v12/data/ValidatingWebhookConfiguration.admissionregistration.k8s.io/aro-admission-controller.aro.openshift.io.yaml:
```diff
@@ -0,0 +1,177 @@
+    - CREATE
+    - UPDATE
+    resources:
+    - replicasets
```
can you condense these hooks per apigroup and version? e.g.
```
- clientConfig:
    caBundle: null
    service:
      name: aro-admission-controller
      namespace: kube-system
      path: /daemonsets
  failurePolicy: Fail
  name: daemonsets.aro-admission-controller.aro.openshift.io
  namespaceSelector: {}
  rules:
  - apiGroups:
    - apps
    apiVersions:
    - v1
    operations:
    - CREATE
    - UPDATE
    resources:
    - daemonsets
    - replicasets
```

Makdaam replied:
Condensed.

jim-minter commented on pkg/sync/v12/sync.go:
```diff
@@ -357,6 +366,48 @@ func (s *sync) writeDB() error {
+		LabelSelector: "openshift.io/component=aro-admission-controller",
+	}
+	s.log.Debug("waiting for admission controller pods")
+	wait.Until(func() {
```
use wait.PollImmediateInfinite

Makdaam replied:
Done.

jim-minter commented on pkg/sync/v12/sync.go:
```diff
@@ -357,6 +366,48 @@ func (s *sync) writeDB() error {
+			if len(pods.Items) == 3 {
+				allReady := true
+				for _, pod := range pods.Items {
+					for _, c := range pod.Status.Conditions {
```
use ready.PodIsReady

Makdaam replied:
Done.

jim-minter commented on pkg/util/roundtrippers/retrying.go:
```diff
@@ -95,16 +95,17 @@ func NewRetryingRoundTripper(log *logrus.Entry, location string, privateEndpoint
 }
 
 func (rt *RetryingRoundTripper) RoundTrip(req *http.Request) (resp *http.Response, err error) {
+	cancel := make(chan struct{})
```
this change doesn't look right - why are you making it?

Makdaam replied:
Reverted.

Makdaam commented on pkg/arm/v12/data/master-startup.sh:
```diff
@@ -71,6 +71,9 @@ for dst in tcp,8444; do
 done
 
+# accepting 8443 for the admission controller
+iptables -A OS_FIREWALL_ALLOW -p tcp -m state --state NEW -m tcp --dport 8443 -j ACCEPT -w
```
Admission controller TCP port is available from within the vnet. Added TLS auth to remedy this situation.

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+	mux.HandleFunc("/healthz/ready", ac.handleHealthz)
+
+	ac.log.Print("Aro Admission Controller starting.")
+	err := http.ListenAndServeTLS(":8443", "/etc/aro-admission-controller/aro-admission-controller.crt", "/etc/aro-admission-controller/aro-admission-controller.key", mux)
```
Done.

Makdaam commented on pkg/config/v12/generate.go:
```diff
@@ -210,6 +210,16 @@ func (g *simpleGenerator) Generate(template *pluginapi.Config, setVersionFields
 			key:  &c.Certificates.ServiceCatalogServer.Key,
 			cert: &c.Certificates.ServiceCatalogServer.Cert,
 		},
+		{
```
Done.

Makdaam commented on pkg/startup/v12/data/master/etc/aro-admission-controller/aro-admission-controller.yaml:
```diff
@@ -0,0 +1,4 @@
+whitelist:
```
Simplified regexes.

Makdaam commented on pkg/startup/v12/data/master/etc/origin/master/master-config.yaml:
```diff
@@ -1,5 +1,15 @@
 admissionConfig:
   pluginConfig:
+    DenyEscalatingExec:
```
Created issue to track the e2e #2078 .

Makdaam commented on pkg/startup/v12/data/master/etc/origin/master/master-config.yaml:
```diff
@@ -1,5 +1,15 @@
+      configuration:
+        apiVersion: v1
+        disable: false
+        kind: DefaultAdmissionConfig
```
Done.

Makdaam commented on pkg/sync/v12/data/SecurityContextConstraints.security.openshift.io/anyuid.yaml:
```diff
@@ -17,6 +17,7 @@ metadata:
   annotations:
     kubernetes.io/description: anyuid provides all features of the restricted SCC
       but allows users to run with any UID and any GID.
+    openshift.io/reconcile-protect: "true"
```
Without this annotation if the customer adds a user or group to the SCC, then it will get reverted in the next Sync pod run.

It's still marked as owned by Sync pod, but not reconciled.

Makdaam replied:
Removing after consultation. openshift.io/reconcile-protect: "true" should be applied by the customer when they decide to modify an SCC

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+func (ac *admissionController) getAdmissionReviewRequest(r *http.Request) (req *admissionv1beta1.AdmissionRequest, errorcode int) {
+	ac.log.Printf("New review request %s", r.RequestURI)
```
Changed most logs to debug level.

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+	Whitelist []string `json:"whitelist"`
+}
+
+func (c *admissionControllerConfig) loadConfig(log *logrus.Entry) *admissionControllerConfig {
```
Changed to error, no more fatal errors inside.

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+	if err != nil {
+		return nil, http.StatusBadRequest
+	}
+	req = reviewIncoming.Request
```
I still need some data from the review request, and it's different data (operation in SCC validation, requesting user in Pod validation). This would either make the getAdmissionReviewRequest function code path specific or return unnecessary values.

Makdaam replied:
Explained on call.
Done.

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	scc := o.(*security.SecurityContextConstraints)
```
Added.

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+	client            internalclientset.Interface
+	restricted        *security.SecurityContextConstraints
+	whitelistedImages []*regexp.Regexp
+	protectedSCCs     map[string]security.SecurityContextConstraints
```
Renamed.

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+
+// verifySCC makes sure that nothing besides additional users or groups are
+// different between the SCC and an SCCTemplate.
+func (ac *admissionController) verifySCC(scc security.SecurityContextConstraints, sccTemplate security.SecurityContextConstraints) errors.Aggregate {
```
Done.

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+func (ac *admissionController) verifySCC(scc security.SecurityContextConstraints, sccTemplate security.SecurityContextConstraints) errors.Aggregate {
+	var errs []error
+	//checking ObjectMeta
+	labels := scc.ObjectMeta.GetLabels()
```
Done.

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+	//Allow only if the new Groups are a superset of the template Groups
+	for _, templateGroup := range sccTemplate.Groups {
+		found := false
+		for _, sccGroup := range scc.Groups {
```
added contains function ([]string, string) since Groups and Users are []string

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+		}
+	}
+	//ignore Users and Groups in further comparison
+	localSccTemplate := sccTemplate.DeepCopy()
```
Done.

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+	localSccTemplate.Users = []string{}
+	localSccTemplate.Groups = []string{}
+	//added only to remove function side effects
+	localScc := scc.DeepCopy()
```
Done.

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+	codec      = legacyscheme.Codecs.CodecForVersions(nil, serializer, nil, runtime.InternalGroupVersioner)
+)
+
+func toInt32Ptr(i int32) *int32 {
```
fixed.

Makdaam commented on pkg/entrypoint/admissioncontroller/scc.go:
```diff
@@ -0,0 +1,379 @@
+	}
+}
+
+func (ac *admissionController) InitProtectedSCCs() map[string]security.SecurityContextConstraints {
```
GetBootstrapSecurityContextConstraints doesn't contain some users (and I'm also guessing groups) assigned to SCCs when compared to what the sync pod applies.

What minimums do we want to enforce?

jim-minter replied:
ACTION: open follow-up - we should get these from the sync interface not duplicate them here.

Makdaam replied:
#2083 created to track this

jim-minter commented:
@Makdaam please rebase

Makdaam commented:
Rebased and switched from v12 to v13 in the code.

Makdaam commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -0,0 +1,217 @@
+func (c *admissionControllerConfig) validate() error {
+	var err error
+	for _, w := range c.Whitelist {
+		_, err = regexp.Compile(w)
```
Done. Regexps are compiled during config loading.

jim-minter commented:
/test e2e-create-20191027-private

jim-minter commented:
/retest

jim-minter commented:
@asalkeld @mjudeikis this is pretty much mergeable now.  Please can you take the time to review your time on Monday so that I can get it merged my time on Monday?

Handy review link: https://github.com/openshift/openshift-azure/pull/2023/files/25e0af4bf8f1c6c03118e7cdb320663f04b28cea..eb0a5e49694c515b183bb7e3fed1d6661d75c4fd?utf8=%E2%9C%93&diff=unified&w=1

jim-minter commented on pkg/arm/v13/data/master-startup.sh:
```diff
@@ -65,7 +65,8 @@ if ! grep /var/lib/etcd /etc/fstab; then
 
-for dst in tcp,8444; do
+# accepting 8443 for the admission controller
+for dst in tcp,8443 tcp,8444; do
```
:8443 is accessible from within the cluster but is protected by TLS client certificate.  It is not in the NSG so is not accessible from outside the cluster.

jim-minter commented on pkg/cluster/kubeclient/roundtrip.go:
```diff
@@ -46,5 +49,16 @@ func NewRestConfig(log *logrus.Entry, config *v1.Config, cs *api.OpenShiftManage
 	restconfig.WrapTransport = roundtrippers.NewRetryingRoundTripper(log, cs.Location, cs.Properties.NetworkProfile.PrivateEndpoint, disableKeepAlives)
 
+	if cs.Properties.NetworkProfile.PrivateEndpoint != nil {
+		restconfig.Dial = func(ctx context.Context, network, addr string) (net.Conn, error) {
```
This code moved as a knock-on effect of the revendor to origin's kubernetes tree.

jim-minter commented on pkg/sync/v10/client.go:
```diff
@@ -7,12 +7,11 @@ import (
 
 	"github.com/sirupsen/logrus"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/apimachinery/pkg/api/meta"
```
Changes to pkg/sync/v* (except v13) are knock-on effects of the revendor to origin's kubernetes tree.

jim-minter commented on pkg/sync/v13/data/ClusterRole.rbac.authorization.k8s.io/customer-admin-cluster.yaml:
```diff
@@ -101,13 +101,18 @@ rules:
   - get
   - list
   - watch
+  - update
```
customer admins are now able to edit SCCs

jim-minter commented on test/clients/openshift/strimzi.go:
```diff
@@ -5,8 +5,8 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"k8s.io/client-go/discovery"
```
more rebase-related changes

jim-minter commented:
note to reviewers: the basic idea is that customer admins can:
* add user/group assignments to system SCCs but otherwise not modify them
* do anything they like with non-system SCCs
* create pods bounded by the usual OpenShift SCCs PLUS the new ARO rule that either:
  * a pod with any privileged containers must be in a system namespace (hence not user-controlled), or
  * all privileged containers in a pod must be whitelisted by ARO config.

jim-minter commented:
/retest

mjudeikis commented on cmd/azure/azure.go:
```diff
@@ -17,7 +19,7 @@ var gitCommit = "unknown"
 func main() {
 	if err := run(); err != nil {
-		panic(err)
+		os.Exit(1)
```
```log.Error(err)
os.Exit(1)``` Or is it intentional to drop error message?

jim-minter replied:
Yes - cobra already prints it out, we were duplicating it.

mjudeikis commented on pkg/entrypoint/admissioncontroller/admissioncontroller.go:
```diff
@@ -1,277 +1,211 @@
-	if r.Header.Get("Content-Type") != "application/json" {
-		return nil, http.StatusUnsupportedMediaType
+func newAdmissionController(ctx context.Context, log *logrus.Entry, configFile string) (*admissionController, error) {
+	/* #nosec - does this tool actually provide any value? */
```
+1

mjudeikis commented on pkg/entrypoint/admissioncontroller/admissioncontroller_test.go:
```diff
@@ -2,1280 +2,188 @@ package admissioncontroller
 
```
Is the name really represents what it is doing? It is confusing. Maybe we should move it to `test/util` or smth and use it in the test where we need it? Or all the methods are local and we dont want to expose them?

jim-minter replied:
This is all local (likely to reuse in v4 though).

mjudeikis commented on pkg/entrypoint/admissioncontroller/podwhitelist.go:
```diff
@@ -5,161 +5,168 @@ import (
+}
+
+func (ac *admissionController) validateWhitelistRequest(req *admission.AdmissionRequest) (field.ErrorList, error) {
+	if strings.HasPrefix(req.Namespace, "openshift-") ||
```
So we relay here on OCP admission controller to make sure users cant create `openshift-myproject` namespace where.  Is `os adm new-project` is not available to users currently?

jim-minter replied:
It should be the case that ARO users can't create *namespaces*, only *projects*.  The project request infrastructure protects against creation of projects with reserved names.

mjudeikis commented:
/retest

mjudeikis commented:
Apart from few nits lgtm. Tests become very complicated but it is hard to avoid this :/

mjudeikis commented:
/lgtm
